### PR TITLE
SQSERVICES-1162-backend-2nd-auth-factor-for-login

### PR DESCRIPTION
**Pull request by battermann**
https://wearezeta.atlassian.net/browse/SQSERVICES-1162

## Todo

- [x] Implement `sendEmail` (currently blocked by https://wearezeta.atlassian.net/browse/SQSERVICES-1307)

## Checklist

 - [x] The **PR Title** explains the impact of the change.
 - [x] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] **changelog.d** contains the following bits of information ([details](https://github.com/wireapp/wire-server/blob/develop/docs/developer/changelog.md)):
   - [x] A file with the changelog entry in one or more suitable sub-sections. The sub-sections are marked by directories inside `changelog.d`.

### CODE REVIEW

File: services/brig/src/Brig/API/Public.hs [hunk 1]
```diff
@@ -1065,10 +1067,43 @@ activate (Public.Activate tgt code dryrun)
...
+          gen <- Code.mk6DigitGen $ Code.ForEmail email
+          mbPendingCode <- lift $ Code.lookup (Code.genKey gen) (scope action)
+          case mbPendingCode of
+            Nothing -> do
+              code <-
+                Code.generate
+                  gen
+                  (scope action)
+                  (Code.Retries 3)
+                  (Code.Timeout (10 * 60)) -- 10 minutes
```
Comment (fisx) [inline services/brig/src/Brig/API/Public.hs:1094]: this needs to be configurable in brig.yaml.  can we use [`setActivationTimeout`](https://github.com/wireapp/wire-server/blob/8281461864f623d78d78ef82a27390a5a06de49b/services/brig/src/Brig/Options.hs#L424)?

File: services/brig/deb/opt/brig/templates/en/user/email/verification-login.txt [hunk 2]
```diff
@@ -0,0 +1,16 @@
+[${brand_logo}]
```
Comment (jschaul) [inline services/brig/deb/opt/brig/templates/en/user/email/verification-login.txt:1]: A corresponding html email should also be included; and ideally through adding the email contents to https://github.com/wireapp/wire-emails first, then re-importing through fetch.py (that would include doing a little bit of maintenance work also)

File: services/brig/src/Brig/API/Public.hs [hunk 3]
```diff
@@ -1063,12 +1065,47 @@ activate (Public.Activate tgt code dryrun)
...
--- Verification
-
-sendVerificationCode :: (Handler r) ()
-sendVerificationCode =
-  case Public.TeamFeatureSndFPasswordChallengeNotImplemented of
-    _ -> pure ()
+sendVerificationCode :: Public.SendVerificationCode -> (Handler r) ()
+sendVerificationCode req = do
+  let action = Public.svcAction req
+  -- TODO(leif): guard as long as not all of the actions are implemented
```
Comment (jschaul) [inline services/brig/src/Brig/API/Public.hs:1071]: I suppose you plan on working on these TODOs still?

File: services/brig/src/Brig/API/Public.hs [hunk 4]
```diff
@@ -1063,12 +1065,47 @@ activate (Public.Activate tgt code dryrun)
...
+      featureStatus <- getFeatureStatus mbAccount
+      case (mbUserId, featureStatus) of
+        (Just userId, Public.TeamFeatureEnabled) -> do
+          gen <- Code.mk6DigitGen $ Code.ForEmail email
+          mbPendingCode <- lift $ Code.lookup (Code.genKey gen) (scope action)
+          case mbPendingCode of
+            Nothing -> do
+              Timeout timeout <- setActivationTimeout <$> view settings
+              code <-
+                Code.generate
```
Comment (jschaul) [inline services/brig/src/Brig/API/Public.hs:1088]: This looks like duplication; similar code exists for password reset, account deletion, and provider creation (grep e.g. for `genKey`). Can that be refactored so we don't introduce very similar logic one more time?

File: services/brig/src/Brig/User/Auth.hs [hunk 5]
```diff
@@ -125,7 +122,35 @@ login (PasswordLogin li pw label _) typ = do
...
+            Nothing -> loginFailedWith LoginCodeRequired u
+            Just c ->
+              case mbEmail of
+                Nothing -> loginFailedNoEmail u
+                Just email -> do
+                  key <- Code.genKey <$> (Code.mk6DigitGen $ Code.ForEmail email)
+                  codeValid <- isJust <$> Code.verify key Code.AccountLogin c
+                  if codeValid
+                    then pure ()
+                    else loginFailedWith LoginNoPendingCode u
```
Comment (jschaul) [inline services/brig/src/Brig/User/Auth.hs:147]: aesthetics: can this multiple-intended case/if chain be flattened out a little for readability? E.g. through the use of `ifNothing` or similar, or through the use of a `case (x, y, z) of ...` pattern ?

File: services/brig/src/Brig/User/Email.hs [hunk 6]
```diff
@@ -196,6 +203,30 @@ renderVerificationMail VerificationEmail {..} VerificationEmailTemplate {..} bra
...
+renderLoginVerificationMail :: Email -> Code.Key -> Code.Value -> VerificationEmailTemplate -> TemplateBranding -> Mail
+renderLoginVerificationMail email key value VerificationEmailTemplate {..} branding =
+  (emptyMail from)
+    { mailTo = [to],
+      -- To make automated processing possible, the activation code is also added to
+      -- headers. {#RefActivationEmailHeaders}
+      mailHeaders =
+        [ ("Subject", toStrict subj),
+          ("X-Zeta-Purpose", "VerificationLogin"),
+          ("X-Zeta-Key", Ascii.toText (fromRange (Code.asciiKey key))),
```
Comment (jschaul) [inline services/brig/src/Brig/User/Email.hs:215]: why is the key added to the headers but not the body of the request?

File: services/brig/src/Brig/User/Template.hs [hunk 7]
```diff
@@ -217,6 +218,13 @@ loadUserTemplates o = readLocalesDir defLocale templateDir "user" $ \fp ->
             <*> pure emailSender
             <*> readText fp "email/sender.txt"
         )
+    <*> ( VerificationEmailTemplate activationUrl
+            <$> readTemplate fp "email/verification-subject.txt"
+            <*> readTemplate fp "email/verification-login.txt"
+            <*> readTemplate fp "email/verification.html"
```
Comment (jschaul) [inline services/brig/src/Brig/User/Template.hs:224]: Isn't the wording a bit different in the txt mail from the html mail? Or what was the reason for introducing a new txt mail but not a corresponding html mail?

File: services/brig/test/integration/API/User/Auth.hs [hunk 8]
```diff
@@ -361,6 +368,28 @@ testSendLoginCode brig = do
   let _timeout = fromLoginCodeTimeout <$> responseJsonMaybe rsp2
   liftIO $ assertEqual "timeout" (Just (Code.Timeout 600)) _timeout
 
+testLoginVerify6DigitEmailCode :: Brig -> Galley -> DB.ClientState -> Http ()
+testLoginVerify6DigitEmailCode brig galley db = do
```
Comment (jschaul) [inline services/brig/test/integration/API/User/Auth.hs:372]: TODO add tests for: 
* testing login fails with a wrong code
* testing login fails with an expired code

File: services/brig/src/Brig/Code.hs [hunk 9]
```diff
@@ -160,13 +161,25 @@ mkGen cfor = liftIO $ do
       let key = mkKey d (Text.encodeUtf8 (emailKeyUniq (mkEmailKey e)))
           val = Value . unsafeRange . Ascii.encodeBase64Url <$> randBytes 15
        in Gen cfor key val
```
Comment (fisx) [inline services/brig/src/Brig/Code.hs:163]: Maybe for symmetry you can factor this out into a function `genEmailLinkGen` next to `mk6DigitGen'`?

I thought I could restructure this to look easier, but now I think I was wrong, it looks mostly sane :+1:   (except for the convoluted requirements due to our lack of understanding of the code ealier that we don't want to touch any more).

File: services/brig/test/integration/API/User/Auth.hs [hunk 10]
```diff
@@ -361,6 +368,28 @@ testSendLoginCode brig = do
...
+  (u, tid) <- createUserWithTeam' brig
+  let checkLogin body status = login brig body PersistentCookie !!! const status === statusCode
```
Comment (fisx) [inline services/brig/test/integration/API/User/Auth.hs:374]: Cover wrong verification code, and test the different error labels?

File: services/brig/test/integration/API/User/Auth.hs [hunk 11]
```diff
@@ -361,6 +368,28 @@ testSendLoginCode brig = do
...
+  Just vcode <- lookupCode db gen Code.AccountLogin
+
+  let loginMissingCode = PasswordLogin (LoginByEmail email) defPassword (Just defCookieLabel) Nothing
+  checkLoginFails loginMissingCode
+
+  let loginWithCode = PasswordLogin (LoginByEmail email) defPassword (Just defCookieLabel) (Just $ Code.codeValue vcode)
+  checkLoginSucceeds loginWithCode
+
+lookupCode :: MonadIO m => DB.ClientState -> Code.Gen -> Code.Scope -> m (Maybe Code.Code)
+lookupCode db gen = liftIO . DB.runClient db . Code.lookup (Code.genKey gen)
```
Comment (fisx) [inline services/brig/test/integration/API/User/Auth.hs:391]: This shouldn't take a `Gen`, but only a key.  And isn't there a function that can convert an email into a key without having to go through generating a `Gen` and then throwing away the bogus code?

File: services/brig/src/Brig/API/Public.hs [hunk 12]
```diff
@@ -1063,12 +1065,47 @@ activate (Public.Activate tgt code dryrun)
...
+                  gen
+                  (scope action)
+                  (Code.Retries 3)
+                  (Code.Timeout timeout)
+                  (Just (toUUID userId))
+              Code.insert code
+              let mbLocale = Public.userLocale <$> accountUser <$> mbAccount
+              lift $ sendMail email (Code.codeKey code) (Code.codeValue code) mbLocale action
+            Just _ -> pure ()
+        _ -> pure ()
```
Comment (fisx) [inline services/brig/src/Brig/API/Public.hs:1098]: I think it's ok to throw an error if the action is invalid.  We document all valid actions in the swagger, and with that error we don't reveal anything beyond that.

File: services/brig/src/Brig/API/Public.hs [hunk 13]
```diff
@@ -1063,12 +1065,47 @@ activate (Public.Activate tgt code dryrun)
...
+  where
+    scope = \case
```
Comment (fisx) [inline services/brig/src/Brig/API/Public.hs:1100]: If you make `scope` a top-level definition, then this is the only place we'd have to revisit every time a new `svcAction` constructor is added, right?

File: services/brig/src/Brig/API/Public.hs [hunk 14]
```diff
@@ -1063,12 +1065,47 @@ activate (Public.Activate tgt code dryrun)
...
+      Public.GenerateScimToken -> error "not implemented (not reachable)" -- TODO(leif): implement
+      Public.Login -> Code.AccountLogin
+    sendMail email key value mbLocale = \case
```
Comment (fisx) [inline services/brig/src/Brig/API/Public.hs:1103]: Can you put empty lines and type signatures here?

File: services/brig/src/Brig/API/Public.hs [hunk 15]
```diff
@@ -1063,12 +1065,47 @@ activate (Public.Activate tgt code dryrun)
...
+      Public.GenerateScimToken -> error "not implemented (not reachable)" -- TODO(leif): implement
+      Public.Login -> sendLoginVerificationMail email key value mbLocale
```
Comment (fisx) [inline services/brig/src/Brig/API/Public.hs:1105]: We probably want to send the `svcAction` to this template as well so we can tell the email recipient what they are about to do.

File: services/brig/src/Brig/API/Public.hs [hunk 16]
```diff
@@ -1063,12 +1065,47 @@ activate (Public.Activate tgt code dryrun)
...
+  case action of
+    Public.GenerateScimToken -> pure ()
+    Public.Login -> do
+      let email = Public.svcEmail req
+      mbUserId <- lift $ UserKey.lookupKey $ UserKey.userEmailKey email
+      mbAccount <- lift $ join <$> Data.lookupAccount `traverse` mbUserId
+      featureStatus <- getFeatureStatus mbAccount
+      case (mbUserId, featureStatus) of
+        (Just userId, Public.TeamFeatureEnabled) -> do
+          gen <- Code.mk6DigitGen $ Code.ForEmail email
```
Comment (fisx) [inline services/brig/src/Brig/API/Public.hs:1082]: Same here as in the test: it's weird that we'd have to create a full `gen` just to get the key.  But if it was like that before, maybe we shouldn't touch it.

File: services/brig/src/Brig/User/Template.hs (hunk 7 above)
Comment (fisx) [reply services/brig/src/Brig/User/Template.hs:224]: Yes (we found that independently in a phone call).

File: services/brig/src/Brig/User/Auth.hs [hunk 17]
```diff
@@ -125,7 +122,35 @@ login (PasswordLogin li pw label _) typ = do
...
+    getEmailAndTeamId :: UserId -> ExceptT LoginError (AppIO r) (Maybe Email, Maybe TeamId)
+    getEmailAndTeamId u = lift $ do
+      mbAccount <- Data.lookupAccount u
+      pure (join $ userEmail <$> accountUser <$> mbAccount, join $ userTeam <$> accountUser <$> mbAccount)
+    loginFailedNoEmail :: UserId -> ExceptT LoginError (AppIO r) ()
+    loginFailedNoEmail = loginFailed
 login (SmsLogin phone code label) typ = do
```
[the remark this answers is not in the document]
Comment (battermann) [reply services/brig/src/Brig/User/Auth.hs:154]: ormolu doesn't like it

File: services/brig/src/Brig/API/Public.hs (hunk 16 above)
Comment (battermann) [reply services/brig/src/Brig/API/Public.hs:1081]: I'd agree, but we need the `gen` further down.

File: services/brig/src/Brig/Options.hs [hunk 18]
```diff
@@ -459,6 +460,9 @@ data Opts = Opts
 data Settings = Settings
   { -- | Activation timeout, in seconds
     setActivationTimeout :: !Timeout,
+    -- | Default verification code timeout, in seconds
+    -- use `setVerificationTimeout` as the getter function which always provides a default value
+    setVerificationTimeoutInternal :: !(Maybe Code.Timeout),
```
Comment (fisx) [inline services/brig/src/Brig/Options.hs:465]:
```suggestion
    setVerificationCodeTimeoutInternal :: !(Maybe Code.Timeout),
```

File: libs/wire-api/src/Wire/API/Routes/Public/Brig.hs [hunk 19]
```diff
@@ -174,7 +174,9 @@ type UserAPI =
       )
     :<|> Named
            "send-verification-code"
-           ( Summary "Send a verification code to a given email address."
+           -- FUTUREWORK(leif): Update description and error response once all actions are implemented
+           ( Summary "Send a verification code to a given email address. It's currently not implemented for all available actions. If a non-implemented action is give, this endpoint will return with a 400 error."
```
[the remark this answers is not in the document]
Comment (battermann) [reply libs/wire-api/src/Wire/API/Routes/Public/Brig.hs:180]: yes, in this PR it is. it's removed in one of the next ones.

File: libs/wire-api/test/golden/Test/Wire/API/Golden/Generated/Login_user.hs [hunk 20]
```diff
@@ -56,7 +58,7 @@ testObject_Login_user_3 =
...
     )
     (Just (CookieLabel {cookieLabelText = "\1082362\66362>XC"}))
-    (Just (ActivationCode "123456"))
+    (Just (Value {asciiValue = (unsafeRange ((fromRight undefined (validate ("RcplMOQiGa-JY")))))}))
```
[the remark this answers is not in the document]
Comment (fisx) [reply libs/wire-api/test/golden/Test/Wire/API/Golden/Generated/Login_user.hs:61]: the email contains the code, both are allowed.

File: services/brig/src/Brig/Code.hs [hunk 21]
```diff
@@ -150,23 +152,45 @@ data Gen = Gen
...
 
+mkKey :: MonadIO m => CodeFor -> m Key
+mkKey cfor = liftIO $ do
+  Just sha256 <- getDigestByName "SHA256"
+  let uniqueK = case cfor of
+        ForEmail e -> emailKeyUniq (mkEmailKey e)
+        ForPhone p -> phoneKeyUniq (mkPhoneKey p)
+  pure $ mkKey' sha256 (Text.encodeUtf8 uniqueK)
+
 -- | Initialise a 'Code' 'Gen'erator for a given natural key.
```
Comment (fisx) [inline services/brig/src/Brig/Code.hs:163]:
```suggestion
-- | Initialise a 'Code' 'Gen'erator for a given natural key.  This generates a link for emails and a 6-digit code for phone.  See also: `mk6DigitGen`.
```

File: services/brig/src/Brig/Code.hs [hunk 22]
```diff
@@ -150,23 +152,45 @@ data Gen = Gen
...
-      let key = mkKey d (Text.encodeUtf8 (phoneKeyUniq (mkPhoneKey p)))
-          val =
-            Value . unsafeRange . Ascii.unsafeFromText . Text.pack . printf "%06d"
-              <$> randIntegerZeroToNMinusOne (10 ^ (6 :: Int))
-       in Gen cfor key val
-    mkKey d = Key . unsafeRange . Ascii.encodeBase64Url . BS.take 15 . digestBS d
+    initGen d (ForEmail e) = mkEmailLinkGen e d
+    initGen d _ = mk6DigitGen' cfor d
+
+mk6DigitGen :: MonadIO m => CodeFor -> m Gen
```
Comment (fisx) [inline services/brig/src/Brig/Code.hs:172]:
```suggestion
-- | Initialise a 'Code' 'Gen'erator for a given natural key.  This generates a 6-digit code, matter whether it is sent to a phone or to an email address.  See also: `mkGen`.
mk6DigitGen :: MonadIO m => CodeFor -> m Gen
```

**Review by fisx — APPROVED** [review]

File: services/brig/src/Brig/User/Auth.hs [hunk 23]
```diff
@@ -125,7 +122,33 @@ login (PasswordLogin li pw label _) typ = do
...
+        mbFeatureEnabled <- Intra.getVerificationCodeEnabled `traverse` mbTeamId
+        pure $ fromMaybe (Public.tfwoapsStatus Public.defaultTeamFeatureSndFactorPasswordChallengeStatus == Public.TeamFeatureEnabled) mbFeatureEnabled
+      when featureEnabled $ do
+        case mbCode of
+          Nothing -> loginFailedWith LoginCodeRequired u
+          Just c ->
+            case mbEmail of
+              Nothing -> loginFailedNoEmail u
+              Just email -> do
+                key <- Code.genKey <$> Code.mk6DigitGen (Code.ForEmail email)
```
Comment (fisx) [inline services/brig/src/Brig/User/Auth.hs:141]: I think there is a refactoring hidden in here that makes it unnecessary to generate another code just to throw it away and keep the key, but I won't insist.  :)  perhaps add a FUTUREWORK?

File: services/brig/src/Brig/Code.hs [hunk 24]
```diff
@@ -150,23 +152,45 @@ data Gen = Gen
     genValue :: IO Value
   }
 
+mkKey :: MonadIO m => CodeFor -> m Key
+mkKey cfor = liftIO $ do
```
[the remark this answers is not in the document]
Comment (battermann) [reply services/brig/src/Brig/Code.hs:156]: I think it is used in `Galley.Cassandra.Code`

File: services/brig/test/integration/API/User/Auth.hs [hunk 25]
```diff
@@ -361,6 +375,91 @@ testSendLoginCode brig = do
...
+  (u, tid) <- createUserWithTeam' brig
+  let Just email = userEmail u
+  let checkLoginFails body =
+        login brig body PersistentCookie !!! do
+          const 403 === statusCode
+          const (Just "code-authentication-required") === errorLabel
+
+  Util.setTeamSndFactorPasswordChallenge galley tid Public.TeamFeatureEnabled
+  Util.generateVerificationCode brig (Public.SendVerificationCode Public.Login email)
+  checkLoginFails $ PasswordLogin (LoginByEmail email) defPassword (Just defCookieLabel) Nothing
```
[the remark this answers is not in the document]
Comment (battermann) [reply services/brig/test/integration/API/User/Auth.hs:441]: I'll leave it like this

**Pull request merged**